File: src/bitboard.cc
#include "bitboard.h"

bool GetBit(Bitboard b, int s) {
    return (b & (1ULL << s)) ? 1 : 0;
}

void SetBit(Bitboard &b, Square s) {
    b |= (1ULL << s);
}

void FlipBit(Bitboard &b, Square s) {
    b ^= (1ULL << s);
}

void PopBit(Bitboard &b, Square s) {
    if (GetBit(b, s)) {
        FlipBit(b, s);
    }
}

void PrintBitboard(Bitboard b) {
    for (int i = 0; i < 8; i++) {
        for (int j = 0; j < 8; j++) {
            int square = i * 8 + j;
            printf(" %d ", GetBit(b, square));
        }
        printf("\n");
    }
    printf("\nBitboard: %lu\n\n", b);
}

File: src/julia/bitboard.jl
using Printf

function GetBit(board::Int64, square::Int64)
	return ((board & (~0 << square)) == board) ? 1 : 0
end

function SetBit(board::Int64, square::Int64)
	return board |= (1 << square)
end

function FlipBit(board::Int64, square::Int64)
	return board ^= (1 << square)
end

function PopBit(board::Int64, square::Int64)
	if GetBit(board, square)
		return FlipBit(board, square)
	end
end

function PrintBitboard(board::Int64)
	for i = 1:8
		for j = 1:8
			square = i*8 + j
			print(GetBit(board, square))
		end
		println()
	end
	@printf "Bitboard: %x\n\n" board
end